### ISSUE TITLE
Filter blob by tags using SAS-token in azure CLI fails with authentication error

### ISSUE THREAD

**Issue by Venu-prahalad** [Azure/azure-cli#24171]
Using Azure CLI I am trying to find all the blobs using tag filters. But I am unable to do so. I get the error

“Authentication failure. This may be caused by either invalid account key, connection string or sas token value provided for your storage account.”

Below is the CLI command I tried

`az storage blob filter --tag-filter  "@container= 'containerxx' and Tag1='value'" --account-name stracc1 --sas-token "sv=2021-08-06&st=2022-10-11T07%3A40%3A00Z&se=2022-10-13T07%3A40%3A47Z&sr=c&sp=racwdxltf&sig=mZ4%2F4oXQV3Y4xcBoxxxxxEjJ1zmozb3ikmEddRc%3D"`

I tried to confirm if the sas token provided is valid by running below command and it does execute fine returning a json. This confirms that the SAS token is valid. You can also see I have given Filter tag permission (f) in sas token generated.

az storage blob exists --account-name stracc1 --container-name containerxx --name yy.html --sas-token "sv=2021-08-06&st=2022-10-11T07%3A40%3A00Z&se=2022-10-13T07%3A40%3A47Z&sr=c&sp=racwdxltf&sig=mZ4%2F4oXQV3Y4xcBoxxxxxEjJ1zmozb3ikmEddRc%3D"

Kindly Help

**Comment by yonzhan** [comment]
@evelyn-ys for awareness

**Comment by Imrankhan-MT** [comment]
Hi @Venu-prahalad  I think you have missing "?" in SAS token the Actual SAS token be "?sv=2021-08-06&st=2022-10-11T07%3A40%3A00Z&se=2022-10-13T07%3A40%3A47Z&sr=c&sp=racwdxltf&sig=mZ4%2F4oXQV3Y4xcBoxxxxxEjJ1zmozb3ikmEddRc%3D" but you have given "sv=2021-08-06&st=2022-10-11T07%3A40%3A00Z&se=2022-10-13T07%3A40%3A47Z&sr=c&sp=racwdxltf&sig=mZ4%2F4oXQV3Y4xcBoxxxxxEjJ1zmozb3ikmEddRc%3D" I think this cause an error. 
Please refer this [SO- thread](https://stackoverflow.com/questions/74040507/filter-blob-by-tags-using-sas-token-in-azure-cli-fails-with-authentication-error).

**Comment by VenkatesanSettu-MT** [comment]
I think you have missing the "?" in the SAS-token. The actual SAS-token be "?sv=2021-08-06&st=2022-10-11T07%3A40%3A00Z&se=2022-10-13T07%3A40%3A47Z&sr=c&sp=racwdxltf&sig=mZ4%2F4oXQV3Y4xcBoxxxxxEjJ1zmozb3ikmEddRc%3D" but have given "sv=2021-08-06&st=2022-10-11T07%3A40%3A00Z&se=2022-10-13T07%3A40%3A47Z&sr=c&sp=racwdxltf&sig=mZ4%2F4oXQV3Y4xcBoxxxxxEjJ1zmozb3ikmEddRc%3D".

Please refer this[ SO-thread](https://stackoverflow.com/questions/74040507/filter-blob-by-tags-using-sas-token-in-azure-cli-fails-with-authentication-error)

**Comment by navba-MSFT** [comment]
@Venu-prahalad As rightly pointed out by @VenkatesanSettu-MT you seem to have missed the "?" in your sas token argument.  I was able to run the below command and it worked just fine. Please update the sas token and test your command again and let me know if you still face the same issue. Awaiting your reply. 

**Azure CLI Command:**
>az storage blob filter --account-name MyStorageAccount --tag-filter ""@container"='MyContainer' and "tag1"='value1' and "tag2"='value2'" --sas-token "?sv=2021-06-08&ss=bfqt&srt=sco&sp=rwdlacupiytfx&se=2022-10-14T14:29:41Z&st=2022-10-14T06:29:41Z&spr=https&sig=Sn8uUTNXXXXXXXXXXXXXXXXXBy5w%3D"

**Output:**
[
  {
    "containerName": "indexblob",
    "name": "req.txt",
    "tags": {
      "tag1": "value1",
      "tag2": "value2"
    }
  },
  {
    "containerName": "indexblob",
    "name": "token_dump.txt",
    "tags": {
      "tag1": "value1",
      "tag2": "value2"
    }
  }
]

**Comment by Venu-prahalad** [comment]
@navba-MSFT Thank you for investigating on this issue. I did give a try but in vain. I still get same error. Below is the screenshot of the actual command which I gave a try.

![image](https://user-images.githubusercontent.com/32172428/195795838-a4a84c35-e1ae-430e-a128-adc290085a88.png)

I could see some difference in the SAS token you have given in the message vs the one I could create. I see some extra query string on your SAS (ss, srt.) which is not present in mine. I used storage explorer to generate SAS on the container and gave all permissions available there except "delete version". Granted permissions are Read, Add, Create, Write, Delete, List, Tag and Find.

**Comment by navba-MSFT** [comment]
@Venu-prahalad Thanks for getting back. The reason for your command failure is below:

_**<AuthenticationErrorDetail>The specified signed resource is not allowed for this resource level</AuthenticationErrorDetail>**_

From the error message, it is pretty clear that the failure is due to the parameter: **sr** (signed resource) in your SAS token.

Generating the _Service_ SAS (container level SAS) will only provide you with **sr=c**. 

So the command in my [above comment](https://github.com/Azure/azure-cli/issues/24171#issuecomment-1278548829) was using an _Account_ sas generated at the Storage account level with **srt=sco**.

So ensure that your SAS token has the right resource type (srt) it could be either **sr=sc** or  **sr=so** or **sr=sco**.

For more information, see [this](https://learn.microsoft.com/en-us/rest/api/storageservices/create-account-sas?redirectedfrom=MSDN#specify-the-account-sas-parameters): 

SignedResourceTypes (srt) | Required. Specifies the signed resource types that are accessible with the account SAS.- Service (s): Access to service-level APIs (for example, Get/Set Service Properties, Get Service Stats, List Containers/Queues/Tables/Shares).- Container (c): Access to container-level APIs (for example, Create/Delete Container, Create/Delete Queue, Create/Delete Table, Create/Delete Share, List Blobs/Files and Directories).- Object (o): Access to object-level APIs for blobs, queue messages, table entities, and files (for example, Put Blob, Query Entity, Get Messages, Create File).You can combine values to provide access to more than one resource type. For example, srt=sc specifies access to service and container resources.
-- | --

**Comment by Venu-prahalad** [comment]
@navba-MSFT One thing confuses me. When I can generate Container level SAS with filter of tag permission (f) set, What should I do extra in this command or what other command should I use to filter blobs by tags? I cant use account level SAS/keys for my case in hand.

**Comment by Imrankhan-MT** [comment]
https://stackoverflow.com/questions/74040507/filter-blob-by-tags-using-sas-token-in-azure-cli-fails-with-authentication-error/74051486#74051486 check this updated answer.

If the provided answer justified your requirement, you accept it as answer so that others who encounter the similar issue, it may be useful for them or community members

**Comment by Venu-prahalad** [comment]
@Imrankhan-MT No, that doesn't solve my original problem. I want blob tag filter to work with container level SAS.

**Comment by Venu-prahalad** [comment]
@navba-MSFT As I think more on this, I am thinking its more of a missing feature/gap because from container level SAS generated with filter permissions on it works well with Azure storage REST API and Azure.Storage.Blobs C# library. but it doesnt work well with Az CLI, Powershell cmdlet, storage explorer.

**Comment by navba-MSFT** [comment]
@Venu-prahalad It appears that PowerShell, CLI, and Storage Explorer are calling [Find Blobs by Tag](https://learn.microsoft.com/en-us/rest/api/storageservices/find-blobs-by-tags) instead of [Find Blobs by Tag in Container](https://learn.microsoft.com/en-us/rest/api/storageservices/find-blobs-by-tags-container) for container-level Find Blobs by Tag.  This is causing container-level SAS URLs to not function properly. We are discussing this with the Product owners. I will keep you posted. There is no ETA for the fix.

**Comment by Venu-prahalad** [comment]
@navba-MSFT Thank you. I much appreciate this move. I and my customers will be looking forward to seeing this working at the earliest.

**Comment by navba-MSFT** [comment]
@Venu-prahalad Thanks for your reply. The tentative ETA for the fix release in AzCLI and Powershell is 11/1/2022. Hope this helps.

**Comment by navba-MSFT** [comment]
@Venu-prahalad The fix in AzCLI is released with cli extension:` storage-blob-preview` 0.6.2. So please update your extension by running the below command:

> **az extension update -n storage-blob-preview**

Once it is updated, You can try to run the `az storage blob filter` CLI command and test. Hope this helps.

Ona side note, This fix is also released in AzPowershell in Az 9.1.0 (Az.Storage 5.1.0):
Gallery Module for Azure PowerShell : https://www.powershellgallery.com/packages/Az/9.1.0 
GitHub release: https://github.com/Azure/azure-powershell/releases/tag/v9.1.0-November2022

**Comment by navba-MSFT** [comment]
@Venu-prahalad I hope you had a chance to update your `storage-blob-preview `extension to 0.6.2 version using the below command. 

> **az extension update -n storage-blob-preview**

Then you can run the below command:

**_Az CLI_**
```
az storage blob filter --tag-filter "tag1='value1'" --container-name "MyContainerName" --account-name MyAccountName --sas-token "?sv=2021-08-06&st=2022-11-04T09%3A02%3A20Z&se=2022-11-05T09%3A02%3A20Z&sr=c&sp=racwdxltf&sig=IP0Ia1NPzXXXXXXXXXXXXXX1RE%3D"
```

**Azure Powershell:**
```
$ctx = New-AzStorageContext -StorageAccountName MyAccountName -SasToken "?sv=2021-08-06&st=2022-11-04T09%3A02%3A20Z&se=2022-11-05T09%3A02%3A20Z&sr=c&sp=racwdxltf&sig=IP0Ia1NPzXXXXXXXXXXXXXXXXXX11RE%3D"

$containerName = "MyContainerName"

Get-AzStorageBlobByTag -TagFilterSqlExpression """tag1""='value1'" -Context $ctx -Container $containerName 
```

**Comment by navba-MSFT** [comment]
@Venu-prahalad The action is currently pending on you try the above action plan. Please feel free to reopen this thread, if you need any further assistance. We would be happy to help.

**Comment by Venu-prahalad** [comment]
Hi @navba-MSFT I confirm this is working as expected and the issue is resolved. Thank you very much.
I see same issue with storage explorer. Should I raise another bug with storage explorer team?

**Comment by navba-MSFT** [comment]
@Venu-prahalad 

> Hi @navba-MSFT I confirm this is working as expected and the issue is resolved. Thank you very much. I see same issue with storage explorer. Should I raise another bug with storage explorer team?

You don't have to raise a new issue for Storage Explorer. This issue for Storage explorer has already been created here:
https://github.com/microsoft/AzureStorageExplorer/issues/6184 and its is addressed.

The fix for Storage explorer will be released by _end of November_. See the milestone here:
https://github.com/microsoft/AzureStorageExplorer/milestone/60

**Comment by Venu-prahalad** [comment]
@navba-MSFT I deeply appreciate your quick and detailed responses to the issue and queries. Thanks again.
